README.md
# Simple rest-api written in Clojure

API data processor 


## Running locally

in order to run locally, add Clojure repl,that runs under Leiningen.
you can also check tests under test directory.
to start the service , run in Repl : (-main).

## EndPoints 

This service supports 4 GET endpoints:
1. http://localhost:8080/get-unique-visited-sites?visitor-id=<ID>
2. http://localhost:8080/get-num-of-sessions?site-url=<SITE_URL>
3. http://localhost:8080/get-median-session-length?site-url=<SITE_URL>
4. http://localhost:8080/health-check  
  
## License

Copyright © 2022 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-02-10
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-02-10
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/home-task/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/home-task/compare/0.1.0...0.1.1


test/home_task/core_test.clj
(ns home-task.core-test
  (:require [clojure.test :refer :all]
            [home-task.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


test/home_task/common_test.clj
(ns home-task.common-test
  (:require [clojure.test :refer :all]
            [home-task.utils.common :refer :all]))

(deftest convert-time-to-record-format-test
  (testing "convert epoch time to  "
    (let [time 1347845034
          expected-res "2012-09-17T01:23:54.000Z"]
      (is (= expected-res (convert-time-to-record-format time))))))

(deftest break-time-test
  (testing "break time to json"
    (let [time "2012-09-17T01:23:54.000Z"
          expected-res {:year    "2012"
                        :month   "09"
                        :day     "17"
                        :hours   "01"
                        :minutes "23"
                        :seconds "54.000"}]
      (is (= expected-res (break-time time))))))

(deftest calculate-succisive-test
  (testing "calculate-succisive testing"
    (let [data [{
                 :visitor_id    "111"
                 :site_url      "22"
                 :page_view_url ""
                 :timestamp     {:year    "2012"
                                 :month   "09"
                                 :day     "17"
                                 :hours   "01"
                                 :minutes "23"
                                 :seconds "54.000"

                                 }},
                {
                 :visitor_id    "111"
                 :site_url      "22"
                 :page_view_url ""
                 :timestamp     {:year    "2012"
                                 :month   "09"
                                 :day     "17"
                                 :hours   "01"
                                 :minutes "01"
                                 :seconds "12.000"}}]
          expected [-22.0]]
      (= expected (calculate-succisive data)))))

(deftest aggregate-results-test
  (testing "aggregate-results testing"
    (let [data [-1.0 -3.0 23.0 -7.0]
          expected-res 34.0]
      (= expected-res (aggregate-results data)))))

project.clj
(defproject home-task "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[fmnoise/flow "4.0.0"]
                 [org.clojure/clojure "1.9.0"]
                 [http-kit "2.3.0"]
                 [org.postgresql/postgresql "9.4-1204-jdbc41"]
                 [cheshire "5.10.0"]
                 [honeysql "1.0.444"]
                 [com.mchange/c3p0 "0.9.5.5"]
                 [compojure "1.6.0"]
                 [amalloy/ring-gzip-middleware "0.1.3"]
                 [ring/ring-json "0.4.0"]
                 [mount "0.1.16"]
                 [org.clojure/java.jdbc "0.7.11"]
                 [javax.servlet/servlet-api "2.5"]
                 [nilenso/honeysql-postgres "0.2.6"]
                 [org.clojure/core.async "1.2.603"]
                 [org.clojure/core.match "0.3.0"]
                 [com.taoensso/carmine "2.16.0"]
                 [javax.annotation/javax.annotation-api "1.3.2"]
                 [javax.xml.bind/jaxb-api "2.3.1"]
                 [clj-time "0.15.1"]]
  :repl-options {:init-ns home-task.core})


src/home_task/utils/common.clj
(ns home-task.utils.common
  (:require [clojure.string :as clj-str]
            [clojure.walk :as clj-wlk]
            [clojure.string :as string]
            [clj-time.format]
            [clj-time.coerce]))

(def headers [:visitor_id, :site_url, :page_view_url, :timestamp])
(def group-keys [:visitor_id, :site_url :page_view_url])

;TODO For Supporting large scale input:
; when we want to scale up this code, we should consider parallelize data
; everytime we use map,we should consider using pmap , this will reduce time significantly
; Regarding data : we should store this data in non-rational DB
; that works well with a high volume of data such as Druid and pull data instead of saving data under Resource Folder.
; we can work with a smaller group of data ,such as parallelize in threads every 6 hours to run separately.
(defn load-csv
  "Attempts to open a csv file and complains if the file is not present."
  [file-name]
  (let [file-data (try
                    (slurp file-name)
                    (catch Exception e (println (.getMessage e))))]
    file-data))

(defn parse-csv-response [raw-data]
  "Parse CSV data ,convert it to json format with relevant headers listed above."
  (try
    (let [rows (clj-str/split-lines raw-data)]
      (if (not-empty rows)
        (->> (map (comp #(zipmap headers %) #(clj-str/split % #",")) rows)
             (map #(clj-wlk/keywordize-keys %)))
        rows))
    (catch Exception e
      (print "Error occurred while parsing body data." (ex-data e)))))

(defn get-data [path]
  (-> path
      (load-csv)
      (parse-csv-response)))

(defn convert-time-to-record-format [time]
  (let [time (* time 1000)
        format (clj-time.format/formatters :date-time)]
    (clj-time.format/unparse format (clj-time.coerce/from-long time))))

(defn break-time [date]
  "Break time to json ,time format yyyy-mm-ddTHH:MM:SS.000Z"
  (let [[year month day-info] (string/split date #"-")
        day (string/split day-info #"T")
        [hours minutes seconds] (string/split (last day) #":")]
    {:year    year
     :month   month
     :day     (first day)
     :hours   hours
     :minutes minutes
     :seconds (string/join "" (drop-last seconds))}))

(defn normalize-timestamp [site-data]
  (map #(assoc % :timestamp (-> (:timestamp %)
                                (Integer/parseInt)          ; epoch string to int
                                (convert-time-to-record-format)
                                (break-time))) site-data))

(defn convert-time-to-minutes [record]
  (let [{:keys [hours minutes]} (:timestamp record)
        hours (Float/parseFloat hours)
        minutes (Float/parseFloat minutes)]
    (+ (* hours 60) minutes)))

(defn calculate-succisive [records]
  (let [data (map #(convert-time-to-minutes %) records)]
    (map - (rest data) data)))

(defn filter-successive-values [records]
  "Drop minutes successive values that are bigger than 30."
  (flatten (remove empty? (mapv (fn [values] (filter (fn [value]
                                                        (if (< value 0)
                                                          (< (* value -1) 30.0)
                                                          (< value 30.0))) values)) records))))

(defn aggregate-results [results]
  (reduce (fn [result input]
            (let [input (if (< input 0) (* -1 input) input)]
              (+ result input))) 0 results))

(defn format-number [number]
  "Take 2 digits after decimal."
  (float (/ (int (* 100 number)) 100)))

(defn session-info [data site-url]
  (let [site-data(filter #(= (:site_url %) site-url) data)
        normalized-data (normalize-timestamp site-data)
        grouped-data (group-by #(select-keys % group-keys) normalized-data)
        filtered-data (filter (fn [[_ v]] (> (count v) 1)) grouped-data)
        successive-values (map (fn [[_ v]] (calculate-succisive v)) filtered-data)
        filtered-successive-values (filter-successive-values successive-values)
        aggregated-sum (aggregate-results filtered-successive-values)
        num-of-sessions (count filtered-successive-values)]
    {:aggregated-sum  aggregated-sum
     :num-of-sessions num-of-sessions}))

(defn calculate-median [data site-url]
  (let [{:keys [aggregated-sum num-of-sessions]} (session-info data site-url)]
    (-> (/ aggregated-sum num-of-sessions) format-number)))

(defn unique-visited-sites [data vistor-id]
  (let [visitor-data (filter #(= (:visitor_id %) vistor-id) data)
        grouped-data (group-by :site_url visitor-data)]
    (count grouped-data)))

src/home_task/http/handlers.clj
(ns home-task.http.handlers
  (:require [ring.middleware.json :refer [wrap-json-body wrap-json-response]]
            [cheshire.core :as json]
            [home-task.utils.common :refer [get-data session-info calculate-median unique-visited-sites]]))
(defonce input_1_path "/Users/sallydabbah/simple-api/resources/input_1.csv")
(defonce input_2_path "/Users/sallydabbah/simple-api/resources/input_2.csv")
(defonce input_3_path "/Users/sallydabbah/simple-api/resources/input_3.csv")

(defn http-response [status body]
  {:status  status
   :headers {"Content-Type" "application/json"}
   :body    (json/generate-string body)})

(defn get-num-of-sessions [site-url]
  (let [data (get-data input_3_path)
        num-of-sessions (:num-of-sessions (session-info data site-url))]
    (http-response 200 (str "Num sessions for site " site-url " = " num-of-sessions))))

(defn get-median-session-length [site-url]
  (let [data (get-data input_3_path)
        median-value (calculate-median data site-url)]
    (http-response 200 (str " median session length = " median-value))))

(defn get-unique-visited-sites [visitor-id]
  (let [data (get-data input_3_path)
        num-unique-sites (unique-visited-sites data visitor-id)]
    (http-response 200 (str "Num of unique sites for " visitor-id " = " num-unique-sites))))

src/home_task/http/server.clj
(ns home-task.http.server
  (:require [org.httpkit.server :as s]
            [home-task.http.handlers :as handler]
            [compojure.route :refer [not-found]]
            [mount.core :refer [defstate]]
            [ring.middleware.json :refer [wrap-json-body wrap-json-response]]
            [compojure.core :refer [defroutes routes GET POST PUT DELETE]]
            [ring.middleware.gzip :refer [wrap-gzip]]
            [compojure.handler :refer [site]]))

(defroutes gen-routes
           (GET "/health-check" []
             (handler/http-response 200 "ok"))
           (GET "/get-num-of-sessions" [site-url]
             (handler/get-num-of-sessions site-url))
           (GET "/get-median-session-length" [site-url]
             (handler/get-median-session-length site-url))
           (GET "/get-unique-visited-sites" [visitor-id]
             (handler/get-unique-visited-sites visitor-id)))

(defn- create-routes []
  (try
    (routes gen-routes)
    (catch Exception e
      (print "Exception occurred in http server:" (.getMessage e))
      (handler/http-response 500 (str "ring handler error: " (.getMessage e))))))

(defn- create-server []
  (s/run-server (-> (site (create-routes))
                    (wrap-json-body {:keywords? true})
                    (wrap-gzip))
                {:port 8080 :thread 50}))

(defstate http-server
          :start (create-server))